master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -653,11 +656,33 @@ private[celeborn] class Master(
       override def run(): Unit = {
         statusSystem.handleAppLost(appId, requestId)
         logInfo(s"Removed application $appId")
+        // only leader can clean hdfs dirs
+        if (conf.hasHDFSStorage && !conf.hdfsDir.isEmpty) {
+          cleanExpiredAppDirsOnHDFS()
```

Comment:
This may cost lot as applications may lost frequently in a big cluster, we would better not clean expired app dirs in handleApplicationLost, instead we need do this in timeoutDeadApplications after we handleApplicationLost.  And if this cost lot, I suggest we can cache this list and refresh every x(3) min turn?

Reply:
Sounds reasonable. I'll move the clean logic to `timeoutDeadApplications`. I think listing directories won't cost a lot.

Reply:
> This may cost lot as applications may lost frequently in a big cluster, we would better not clean expired app dirs in handleApplicationLost, instead we need do this in timeoutDeadApplications after we handleApplicationLost. And if this cost lot, I suggest we can cache this list and refresh every x(3) min turn?

I also think it's not a good idea to call ```cleanExpiredAppDirsOnHDFS``` when app lost. IMO we can just use ```forwardMessageThread.scheduleAtFixedRate``` to check like ```checkForWorkerTimeOutTask``` and ```checkForApplicationTimeOutTask```. Also we need to change ```forwardMessageThread``` from single thread to multiple threads.

Reply:
Also, we need to ensure that the leader has replayed all raft logs before cleanup.